parsec/core/mountpoint/winfsp_runner.py

```diff
@@ -154,25 +135,40 @@ async def winfsp_mountpoint_runner(
         # security_timeout=10000,
     )
     try:
-        event_bus.send("mountpoint.starting", mountpoint=mountpoint_path)
+        # Prepare event information
+        event_kwargs = {
+            "mountpoint": mountpoint_path,
+            "workspace_id": workspace_fs.workspace_id,
+            "timestamp": getattr(workspace_fs, "timestamp", None),
+        }
+        event_bus.send("mountpoint.starting", **event_kwargs)
+
+        # Manage drive icon
+        drive_letter, *_ = mountpoint_path.drive
+        with parsec_drive_icon_context(drive_letter):
+
+            # Run fs start in a thread
+            await trio.to_thread.run_sync(fs.start)
+            await _wait_for_winfsp_ready(mountpoint_path)
+
+            # Notify the manager that the mountpoint is ready
+            event_bus.send("mountpoint.started", **event_kwargs)
+            task_status.started(mountpoint_path)
 
-        # Run fs start in a thread, as a cancellable operation
-        # This is because fs.start() might get stuck for while in case of an IRP timeout
-        await trio.to_thread.run_sync(fs.start, cancellable=True)
+            # Start recording `sharing.updated` events
+            with event_bus.waiter_on("sharing.updated") as waiter:
 
-        # Because of reject_irp_prior_to_transact0, the mountpoint isn't ready yet
-        # We have to add a bit of delay here, the tests would fail otherwise
-        # 10 ms is more than enough, although a strict process would be nicer
-        # Still, this is only temporary as avast is working on a fix at the moment
-        # Another way to address this problem would be to migrate to python 3.8,
-        # then use `os.stat` to differentiate between a started and a non-started
-        # file syste.
-        await trio.sleep(0.01)
+                # Loop over `sharing.updated` event
+                while True:
 
-        event_bus.send("mountpoint.started", mountpoint=mountpoint_path)
-        task_status.started(mountpoint_path)
+                    # Restart workspace if necessary
+                    if workspace_fs.is_read_only() != fs.volume_params["read_only_volume"]:
+                        restart = partial(fs.restart, read_only_volume=workspace_fs.is_read_only())
+                        await trio.to_thread.run_sync(restart)
 
-        await trio.sleep_forever()
+                    # Wait and reset waiter
+                    await waiter.wait()
```

Comment:
Shouldn't `waiter.wait()` be first in the loop ?

Reply:
No:
- First we start the mountpoint
- Then we connect to the events

That means we could miss an event in between those two points. I think the following logic prevents race condition in general:
```python
with listen() as event:
    while True:
        if fix_needed():
            fix()
        event.wait_and_reset()
```

Reply:
👍